### DISCUSSION TITLE
Overzealous indexer

### DISCUSSION THREAD

**Discussion by gvl2023** [eclipse-cdt/cdt#304] [Q&A]
I have a makefile project. Makefiles are generated by cmake with unity builds enabled. The project has separate build/source folders, conceptually it is something like this:

```
project
  build
    unity.cpp
    unity.cpp.o
    unity.cpp.o.d
    Makefile
  src
    source1.cpp
    source2.cpp
```

Both _build_ and _src_ are linked to folders outside of workspace. _unity.cpp_ includes both of the source files. Source location is rooted at the project level. Project refresh policy is set to refresh _build_. If I change _source1.cpp_ and build the project it kicks off the indexer and it looks like it then reindexes the whole project. I have the "auto-update index" enabled but It appears there should be no changes to any of the source files under the build folder as a result of building the project so it's not clear to me why the indexer even starts. Can someone help me understand what is happening? Is there a way to enable debug trace or indexer log?

**Comment by jonahgraham** [comment]
The answer to your last part first - have a look at [How to troubleshoot indexing issues](https://github.com/eclipse-cdt/cdt/blob/main/FAQ/Indexer.md)

As for your main question - I am seeing more and more unity builds questions recently and I am not sure if any of the active CDT maintainers have looked at this area yet.

There are also lots of [tracing options](https://github.com/eclipse-cdt/cdt/blob/7c8bb9f00ef0d5e3ff493bc0c944519a3e476da2/core/org.eclipse.cdt.core/.options#L1) in CDT's core bundle. See [FAQ How do I use the platform debug tracing facility](https://wiki.eclipse.org/FAQ_How_do_I_use_the_platform_debug_tracing_facility) (that link may be a little low level, somewhere there is a better document on how to set it up, but I can't find it right now so hopefully that is a good place to start)

>  Source location is rooted at the project level. 

It seems to me that as the unity files have all the sources again that the source location should be set to src so you don't index unity.cpp. 

> Project refresh policy is set to refresh build.

What about workspace refresh? Is it off:

![image](https://user-images.githubusercontent.com/679236/223193844-10968f00-fdf0-47c3-b727-24b7cdfc1171.png)

**Reply by gvl2023** [reply]
Workspace is set to refresh on access only. As for setting the source location to src, I think this will prevent the indexer from using discovered -I and -D values as they are kept with the unity files, or rather folders containing them as this is how I set it up. I'll try to get trace working and make sense of what is happening.

**Reply by jonahgraham** [reply]
Let me know if you need guidance on setting up the tracing options.
